File: src/main/java/earth/terrarium/tempad/common/items/TempadItem.java
package earth.terrarium.tempad.common.items;

import dev.architectury.injectables.annotations.PlatformOnly;
import earth.terrarium.tempad.api.apps.TempadAppApi;
import earth.terrarium.tempad.api.options.FuelOption;
import earth.terrarium.tempad.api.options.FuelOptionsApi;
import earth.terrarium.tempad.api.power.PowerSettings;
import earth.terrarium.tempad.api.power.PowerSettingsApi;
import earth.terrarium.tempad.client.config.TempadClientConfig;
import earth.terrarium.tempad.common.Tempad;
import earth.terrarium.tempad.common.config.CommonConfig;
import earth.terrarium.tempad.api.locations.LocationData;
import earth.terrarium.tempad.common.entity.TimedoorEntity;
import earth.terrarium.tempad.common.registry.TempadRegistry;
import earth.terrarium.tempad.common.utils.ClientUtils;
import earth.terrarium.tempad.common.utils.LookupLocation;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Mth;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class TempadItem extends Item {
    public TempadItem(Properties properties) {
        super(properties);
    }

    @Override
    public @NotNull InteractionResultHolder<ItemStack> use(Level level, Player player, @NotNull InteractionHand interactionHand) {
        ItemStack stack = player.getItemInHand(interactionHand);
        var lookup = new LookupLocation("main", player.getInventory().findSlotMatchingItem(stack));
        if (!player.isShiftKeyDown()) {
            if (level.isClientSide) {
                TempadAppApi.API.getHomePage().openOnClient(player, lookup);
                return InteractionResultHolder.pass(stack);
            } else {
                return InteractionResultHolder.success(stack);
            }
        } else {
            if (level.isClientSide) {
                ClientUtils.openFavorited(lookup);
                return InteractionResultHolder.pass(stack);
            } else {
                return InteractionResultHolder.success(stack);
            }
        }
    }

    public static void summonTimeDoor(LocationData locationData, Player player, int color) {
        TimedoorEntity timedoor = new TimedoorEntity(TempadRegistry.TIMEDOOR_ENTITY.get(), player.level());
        var angle = player.getYHeadRot();
        float angleInRadians = (float) Math.toRadians(angle + 90);
        timedoor.setColor(color);
        timedoor.setTargetLocation(locationData);
        timedoor.setOwner(player.getUUID());
        var position = player.position();
        timedoor.setPos(position.x() + Mth.cos(angleInRadians) * CommonConfig.distanceFromPlayer, position.y(), position.z() + Mth.sin(angleInRadians) * CommonConfig.distanceFromPlayer);
        timedoor.setYRot(angle);
        player.level().addFreshEntity(timedoor);
        timedoor.playSound(Tempad.TIMEDOOR_SOUND.get());
    }

    @Override
    public void appendHoverText(@NotNull ItemStack stack, @Nullable Level level, @NotNull List<Component> components, @NotNull TooltipFlag flag) {
        FuelOption option = FuelOptionsApi.API.findItemOption(stack);
        PowerSettings attachment = PowerSettingsApi.API.get(stack);
        option.addToolTip(stack, attachment, level, components, flag);
    }

    @Override
    public int getBarColor(@NotNull ItemStack $$0) {
        return TempadClientConfig.color;
    }

    @Override
    public boolean isBarVisible(@NotNull ItemStack stack) {
        return FuelOptionsApi.API.findItemOption(stack).isDurabilityBarVisible(stack, PowerSettingsApi.API.get(stack));
    }

    @Override
    public int getBarWidth(@NotNull ItemStack stack) {
        return (int) (FuelOptionsApi.API.findItemOption(stack).getPercentage(stack, PowerSettingsApi.API.get(stack)) * 13);
    }

    @Override
    public void inventoryTick(ItemStack stack, Level level, Entity entity, int slotId, boolean isSelected) {
        super.inventoryTick(stack, level, entity, slotId, isSelected);
        FuelOptionsApi.API.findItemOption(stack).tick(stack, PowerSettingsApi.API.get(stack), entity);
    }

    public boolean shouldCauseReequipAnimation(ItemStack oldStack, ItemStack newStack, boolean slotChanged) {
        return false;
    }
}

File: src/main/kotlin/earth/terrarium/tempad/common/items/TempadItem.kt
package earth.terrarium.tempad.common.items

import earth.terrarium.tempad.api.access.ItemAccessAddress
import earth.terrarium.tempad.api.app.TempadAppRegistry
import earth.terrarium.tempad.api.macro.MacroRegistry
import earth.terrarium.tempad.common.registries.*
import earth.terrarium.tempad.common.utils.contents
import earth.terrarium.tempad.common.utils.ctx
import earth.terrarium.tempad.common.utils.getSlot
import earth.terrarium.tempad.common.utils.stack
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.SlotAccess
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.ClickAction
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level

class TempadItem(props: Properties) : ChrononItem(props) {
    override fun inventoryTick(
        itemStack: ItemStack,
        level: ServerLevel,
        owner: Entity,
        slot: EquipmentSlot?,
    ) {
        super.inventoryTick(itemStack, level, owner, slot)
        if (itemStack.serialNumber == null) {
            itemStack.serialNumber = Component.literal("${itemStack.serialPrefix}-${owner.random.nextInt(1000, 10_000)}-${owner.plainTextName.take(4).uppercase()}")
        }
    }

    override fun use(level: Level, player: Player, hand: InteractionHand): InteractionResult {
        val stack = player.getItemInHand(hand)
        if (level.isClientSide) return InteractionResult.SUCCESS

        val slotId = hand.getSlot(player)
        val access = player.ctx(slotId)
        val accessAddress = ItemAccessAddress(ModItemAccess.inventory, slotId)

        if (player.isShiftKeyDown) {
            MacroRegistry[stack.defaultMacro]?.run(player, access)
        } else {
            (TempadAppRegistry[stack.defaultApp, player, accessAddress]?: TempadAppRegistry[ModApps.teleport, player, accessAddress])!!.openMenu(player as ServerPlayer)
        }

        return InteractionResult.SUCCESS
    }

    override fun overrideStackedOnOther(stack: ItemStack, slot: Slot, action: ClickAction, player: Player): Boolean {
        if (action == ClickAction.SECONDARY) {
            if (stack.twisterEquipped) {
                if (!slot.hasItem()) {
                    stack.twisterEquipped = false
                    slot.contents = ModItems.timeTwister.stack {
                        stack.twisterData?.let(::applyComponents)
                        chrononContent = stack.chrononContentTimeTwister
                        stack.chrononContentTimeTwister = 0
                    }
                    return true
                } else if (slot.contents.item === ModItems.tempad && !slot.contents.twisterEquipped) {
                    stack.twisterEquipped = false
                    slot.contents.twisterEquipped = true
                    slot.contents.chrononContentTimeTwister = stack.chrononContentTimeTwister
                    slot.contents.twisterData = stack.twisterData
                    stack.chrononContentTimeTwister = 0
                    stack.twisterData = null
                    return true
                }
            } else if (slot.contents.item === ModItems.timeTwister) {
                stack.twisterEquipped = true
                stack.chrononContentTimeTwister += slot.contents.chrononContent
                stack.twisterData = slot.contents.componentsPatch
                slot.contents = ItemStack.EMPTY
                return true
            }
        }
        return super.overrideStackedOnOther(stack, slot, action, player)
    }

    override fun overrideOtherStackedOnMe(
        stack: ItemStack,
        other: ItemStack,
        slot: Slot,
        action: ClickAction,
        player: Player,
        access: SlotAccess,
    ): Boolean {
        if (action == ClickAction.SECONDARY && other.item === ModItems.timeTwister && !stack.twisterEquipped) {
            stack.twisterEquipped = true
            stack.chrononContentTimeTwister += other.chrononContent
            stack.twisterData = other.componentsPatch
            access.set(ItemStack.EMPTY)
            return true
        }
        return super.overrideOtherStackedOnMe(stack, other, slot, action, player, access)
    }
}